File: scripts/write_preview.rb
require_relative 'lib.rb'

def write_vertical_spacing
  for _ in 0..1
    set_color :background, :black
    for _ in 0..41
      print ' '
    end
    reset_colors
    puts
  end
end

def write_tribal_decoration(length)
  for column in 0..length - 1
    is_even = column % 2 == 0
    set_color :foreground, is_even ? :yellow : :red
    print is_even ? '≥' : 'v'
  end
end

def write_header
  set_color :background, :black
  print '     '
  write_tribal_decoration 10
  set_color :foreground, :white
  print ' Flamerial '
  write_tribal_decoration 11
  print '     '
  reset_colors
  puts
end

def write_palette(layer)
  set_color :background, :black
  set_color :foreground, :white
  print "     #{layer == :foreground ? '¦\\:' : '¦/:'} "
  for key, _ in $colors.to_a[2..].to_h
    set_color layer, key
    print layer == :foreground ? 'Fla' : '   '
  end
  set_color :background, :black
  set_color :foreground, :white
  print " #{layer == :foreground ? ':/¦' : ':\\¦'}     "
  reset_colors
  puts
end

def write_footer()
  set_color :background, :black
  print '     '
  write_tribal_decoration 32
  print '     '
  reset_colors
  puts
end

write_vertical_spacing
write_header
write_palette :background
write_palette :foreground
write_footer
write_vertical_spacing

File: scripts/write_preview.rs
//! ```cargo
//! [dependencies]
//! anyhow = "1.0.98"
//! crossterm = "0.29.0"
//! heck = "0.5.0"
//! ```

use anyhow::Result;
use crossterm::{
    execute,
    style::{Print, ResetColor, SetBackgroundColor, SetForegroundColor},
};
use std::io::{stdout, Stdout};

include!("lib.rs");

fn write_blank_spacing(stdout: &mut Stdout) -> Result<()> {
    for _ in 0..2 {
        execute!(
            stdout,
            SetBackgroundColor(PALETTE.black().as_crossterm_color()),
        )?;
        for _ in 0..42 {
            execute!(stdout, Print(" "))?;
        }
        execute!(stdout, ResetColor, Print("\n"))?;
    }
    Ok(())
}

fn write_tribal_decoration(stdout: &mut Stdout, length: usize) -> Result<()> {
    for column in 0..length {
        let is_even = column % 2 == 0;
        execute!(
            stdout,
            SetForegroundColor(
                if is_even {
                    PALETTE.yellow()
                } else {
                    PALETTE.red()
                }
                .as_crossterm_color()
            ),
            Print(if is_even { "≥" } else { "v" })
        )?;
    }
    Ok(())
}

fn write_header(stdout: &mut Stdout) -> Result<()> {
    execute!(
        stdout,
        SetBackgroundColor(PALETTE.black().as_crossterm_color()),
        Print("     ")
    )?;
    write_tribal_decoration(stdout, 10)?;
    execute!(
        stdout,
        SetForegroundColor(PALETTE.white().as_crossterm_color()),
        Print(" Flamerial ")
    )?;
    write_tribal_decoration(stdout, 11)?;
    execute!(stdout, Print("     "), ResetColor, Print("\n"))?;
    Ok(())
}

fn write_palette(stdout: &mut Stdout, layer: Layer) -> Result<()> {
    execute!(
        stdout,
        SetBackgroundColor(PALETTE.black().as_crossterm_color()),
        SetForegroundColor(PALETTE.white().as_crossterm_color()),
        Print(format!(
            "     {} ",
            if layer == Layer::Background {
                "¦/:"
            } else {
                "¦\\:"
            }
        ))
    )?;
    for color in PALETTE.preview_colors() {
        if layer == Layer::Background {
            execute!(
                stdout,
                SetBackgroundColor(color.as_crossterm_color()),
                Print("   ")
            )?;
        } else {
            execute!(
                stdout,
                SetForegroundColor(color.as_crossterm_color()),
                Print("Fla")
            )?;
        }
    }
    execute!(
        stdout,
        SetBackgroundColor(PALETTE.black().as_crossterm_color()),
        SetForegroundColor(PALETTE.white().as_crossterm_color()),
        Print(format!(
            " {}     ",
            if layer == Layer::Background {
                ":\\¦"
            } else {
                ":/¦"
            }
        )),
        ResetColor,
        Print("\n")
    )?;
    Ok(())
}

fn write_footer(stdout: &mut Stdout) -> Result<()> {
    execute!(
        stdout,
        SetBackgroundColor(PALETTE.black().as_crossterm_color()),
        Print("     ")
    )?;
    write_tribal_decoration(stdout, 32)?;
    execute!(stdout, Print("     "), ResetColor, Print("\n"))?;
    Ok(())
}

fn main() -> Result<()> {
    let mut stdout = stdout();
    write_blank_spacing(&mut stdout)?;
    write_header(&mut stdout)?;
    write_palette(&mut stdout, Layer::Background)?;
    write_palette(&mut stdout, Layer::Foreground)?;
    write_footer(&mut stdout)?;
    write_blank_spacing(&mut stdout)?;
    Ok(())
}
